examples/megssm/models.py

```diff
@@ -0,0 +1,872 @@
+        else:
+            Y = obs
+            w_s = 1.
+
+        return Y, w_s, fwd_roi_snsr, fwd_src_snsr, snsr_cov, which_roi
+
+
+# TODO: add documentation to all methods
+# TODO: make some methods "private" (leading underscore) if necessary
+class MEGLDS(_MEGModel):
+    """ State-space model for MEG data, as described in "A state-space model of
+        cross-region dynamic connectivity in MEG/EEG", Yang et al., NIPS 2016.
+    """
+
+    def __init__(self, num_roi, timepts, A=None, roi_cov=None, mu0=None, roi_cov_0=None, 
+                 log_sigsq_lst=None, lam0=0., lam1=0., penalty='ridge', 
+                 store_St=True):
+
+        super().__init__()
+
+        set_default = \
+            lambda prm, val, deflt: \
+                self.__setattr__(prm, val.copy() if val is not None else deflt)
+
+        # initialize parameters
+        set_default("A", A,
+                    np.stack([rand_stable(num_roi, maxew=0.7) for _ in range(timepts)],
+                             axis=0))
+        set_default("roi_cov", roi_cov, rand_psd(num_roi))
+        set_default("mu0", mu0, np.zeros(num_roi))
+        set_default("roi_cov_0", roi_cov_0, rand_psd(num_roi))
+        set_default("log_sigsq_lst", log_sigsq_lst,
+                    [np.log(np.random.gamma(2, 1, size=num_roi+1))])
+
+        self.lam0 = lam0
+        self.lam1 = lam1
+
+        if penalty not in ('ridge', 'lasso', 'group-lasso'):
+            raise ValueError('penalty must be one of: ridge, lasso,' \
+                             + ' group-lasso')
+        self._penalty = penalty
+
+        # initialize lists of smoothed estimates
+        self._mus_smooth_lst = None
+        self._sigmas_smooth_lst = None
+        self._sigmas_tnt_smooth_lst = None
+        self._loglik = None
+        self._store_St = bool(store_St)
+
+        # initialize sufficient statistics
+        timepts, num_roi, _ = self.A.shape
+        self._B0 = np.zeros((num_roi, num_roi))
+        self._B1 = np.zeros((timepts-1, num_roi, num_roi))
+        self._B3 = np.zeros((timepts-1, num_roi, num_roi))
+        self._B2 = np.zeros((timepts-1, num_roi, num_roi))
+        self._B4 = list()
+        
+        self._subject_data = dict()
+        
+    def add_subject(self, subject, subject_dir, epochs,labels, fwd, cov):
+        roi_to_src = ROIToSourceMap(fwd, labels) # compute ROI-to-source map
+        scales = {'eeg_scale' : 1, 'mag_scale' : 1, 'grad_scale' : 1} #scale=1 has no effect
+        fwd_src_snsr, fwd_roi_snsr, snsr_cov, epochs = \
+                _scale_sensor_data(epochs, fwd, cov, roi_to_src, **scales)
+                
+        # which_roi = roi_to_src.which_roi # array of len(sources); val = ROI of source
+        # data = epochs._data
+        # data = np.swapaxes(data,-1,-2)
+        # subjectdata = [(data, fwd_roi_snsr, fwd_src_snsr, snsr_cov, which_roi)]
+        
+        sdata = run_pca_on_subject(subject,epochs, fwd, cov, labels)
```

Comment:
... specifically I guess this `run_pca_on_subject` function should take into account channel mismatches between `cov`, `fwd`, and `epochs`